docs/proposals/makefile-improvement-proposal.md

```diff
@@ -0,0 +1,68 @@
+
+## Introduction
+
+The motivation for this proposal is to enable developers working on Argo Workflows to use build tools in a more reproducible way.
+Currently the Makefile is unfortunately too opinionated and as a result is often a blocker when first setting up Argo Workflows locally.
+I believe we should shrink the responsibilities of the Makefile and where possible outsource areas of responsibility to more specialized technology, such
+as Devenv/Nix in the case of dependency management.
+
+## Proposal Specifics
+
+In order to better address reproducibility, it is better to split up the duties the Makefile currently performs into various sub components, that can be assembled in more appropriate technology. One important aspect here is to completely shift the responsibility of dependency management away from the Makefile and into technology such as Nix or Devenv.
+
+### Devenv
+
+#### Benefits of Devenv
+
+- Reproducible build environment
+- Ability to run processes
+
+#### Disadvantages of Devenv
+
+- Huge learning curve to tap into Nix functionality
+- Less documentation
+
+### Nix
+
+#### Benefits of Nix
+
+- Reproducible build environment
+- Direct raw control of various Nix related functionality instead of using Devenv
+- More documentation
+
+#### Disadvantages of Nix
+
+- Huge learning curve
+
+### Recommendation
+
+I suggest that we use Nix over Devenv. I believe that our build environment is unique enough that we will be tapping into Nix anyway, it probably makes sense to directly use Nix in that case.
+
+### Proposal
+
+In order to maximize the benefit we receive from using something like Nix, I suggest that we initially start off with a modest change to the Makefile.
+The first proposal would be to remove out all dependency management code and replace this functionality with Nix, where it is trivially possible. This may not be possible for some go lang related binaries we use, we will retain the Makefile functionality in those cases, at least for a while. Eventually we will migrate more and more of this responsibility away from the Makefile. Following Nix being responsible for all dependency management, we could start to consider moving more of our build system itself into Nix, perhaps it is easiest to start off with UI build as it is relatively painless. However, do note that this is not a requirement, I do not see a problem with the Makefile and the Nix file co-existing, it is more about finding a good balance between the reproducibility we desire and the effort we put into obtaining said reproducibility
+
+### What will Nix get us?
+
+As mentioned previously Nix gets us closer to reproducible build environment's. It should ease significantly the on-boarding process of developers onto the project.
+There have been at least one confirmed developer who wanted to work on Argo Workflows but found the Makefile to be a barrier, it is likely that there are more developers on this boat. With a reproducible build environment, we hope that
```

Comment:
We've had several mentees provide feedback that getting started with the Argo Workflows makefile is challenging and time consuming, so I would say 5+ developers here instead of 1.